hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/pubsub/PubSubMessageSenderProvider.java

```diff
@@ -0,0 +1,143 @@
+                        new ThreadFactoryBuilder().setNameFormat("pubsub-publisher-%d").build()));
+        this.retrySettings = retrySettings();
+        this.batchingSettings = batchingSettings();
+    }
+
+    @Override
+    public MessageSender create(Subscription subscription) {
+        final Publisher publisher = publishersCache.computeIfAbsent(subscription, s -> {
+            final PubSubSenderTarget resolvedTarget = resolver.resolve(s.getEndpoint());
+            return createPublisher(resolvedTarget);
+        });
+        Preconditions.checkNotNull(publisher, "PubSub Publisher cannot be null");
+
+        return new PubSubMessageSender(new PubSubClient(publisher, messageCreator));
+    }
+
+    @Override
+    public Set<String> getSupportedProtocols() {
+        return ImmutableSet.of(SUPPORTED_PROTOCOL);
+    }
+
+    @Override
+    public void start() throws Exception {
+    }
+
+    @Override
+    public void stop() throws Exception {
+        // TODO check that method, because it looks like it is never called across the Hermes codebase
+        publishersCache.values().parallelStream().forEach(p -> {
+            p.shutdown();
+            try {
+                p.awaitTermination(30, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                logger.error("Interrupted termination of the PubSub publisher, it may result in messages loss");
+            }
+        });
+        publishingExecutorProvider.getExecutor().shutdown();
+    }
+
+    private Publisher createPublisher(PubSubSenderTarget resolvedTarget) {
+        try {
+            return Publisher.newBuilder(resolvedTarget.getTopicName())
+                    .setEndpoint(resolvedTarget.getPubSubEndpoint())
+                    .setCredentialsProvider(credentialsProvider.getProvider())
+                    .setRetrySettings(retrySettings)
+                    .setBatchingSettings(batchingSettings)
+                    .setExecutorProvider(publishingExecutorProvider)
+                    .build();
+        } catch (IOException e) {
+            logger.error("Cannot instantiate PubSub Publisher", e);
+            return null;
+        }
+    }
+
+    private BatchingSettings batchingSettings() {
+        long requestBytesThreshold = configFactory.getLongProperty(Configs.PUBSUB_SENDER_REQUEST_BYTES_THRESHOLD);
+        long messageCountBatchSize = configFactory.getLongProperty(Configs.PUBSUB_SENDER_MESSAGE_COUNT_BATCH_SIZE);
+        Duration publishDelayThreshold = Duration.ofMillis(
+                configFactory.getLongProperty(Configs.PUBSUB_SENDER_PUBLISH_DELAY_THRESHOLD));
+
+        return BatchingSettings.newBuilder()
+                .setElementCountThreshold(messageCountBatchSize)
+                .setRequestByteThreshold(requestBytesThreshold)
+                .setDelayThreshold(publishDelayThreshold)
+                .build();
+    }
+
+    private RetrySettings retrySettings() {
```

Comment:
Hermes already has a retry mechanism: [ConsumerMessageSender::retrySending](https://github.com/allegro/hermes/blob/master/hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/ConsumerMessageSender.java#L186-L196). If I understand correctly, by introducing retries at this level we amplify the number of requests sent to the subscriber. In other words, each retry triggered by Hermes will result in X retries triggered by `PubSubClient`.

Can we stay with only retries provided by Hermes?

Reply:
That's reasonable argument but you can neither skip RetrySettings nor create a "no retries at all" object in the Publisher configuration. Even if you create such a retry configuration it is either overridden in the Publisher itself or it throws an exception.
What I can do is:
- set the initial rpc timeout to some value, let's say 10000ms
- set total timeout and max rpc timeout to 10001ms.

Reply:
Or we can set `maxAttempts` to 1. By the default this value is set to 0, and from the publisher code I conclude that it means "retry indefinitely".

Reply:
I missed this setting to be honest. And I've made it simpler. I think only totalTimeoutMs and maxAttempts are needed here. I hardcoded the latter.